minigalaxy/installer.py

```diff
@@ -341,31 +342,61 @@ def compare_directories(dir1, dir2):
     return result
 
 
-def remove_installer(game: Game, installer: str, install_dir: str, keep_installers: bool):
-    error_message = ""
-    installer_directory = os.path.dirname(installer)
-    if not os.path.isdir(installer_directory):
-        error_message = "No installer directory is present: {}".format(installer_directory)
+def remove_installer(game: Game, installer: str, keep_installers_dir: str, keep_installers: bool, file_list=None):
+    installer_dir = os.path.dirname(installer)
+    if not os.path.isdir(installer_dir):
+        error_message = "No installer directory is present: {}".format(installer_dir)
         return error_message
 
-    if keep_installers:
-        keep_dir = os.path.join(install_dir, "installer")
-        keep_dir2 = os.path.join(keep_dir, game.get_install_directory_name())
-        if keep_dir2 == installer_directory:
-            # We are using the keep installer already
-            return error_message
+    installer_root_dirs = [
+        os.path.realpath(DOWNLOAD_DIR),
+        keep_installers_dir
+    ]
 
-        if not compare_directories(installer_directory, keep_dir2):
-            shutil.rmtree(keep_dir2, ignore_errors=True)
-            try:
-                shutil.move(installer_directory, keep_dir2)
-            except Exception as e:
-                error_message = str(e)
-    else:
-        for file in os.listdir(installer_directory):
-            os.remove(os.path.join(installer_directory, file))
+    keep_files = []
+    if keep_installers and file_list:
+        keep_files = file_list
+    elif keep_installers:
+        # assume all support files are named with the same prefix and only differ in ending
+        # we run into this case when installing from local file by ui clicking instead of from the download_finish callback
+        installer_prefix = os.path.splitext(os.path.basename(installer))[0]
+        for f in os.listdir(installer_dir):
+            fullpath = os.path.join(installer_dir, f)
+            if os.path.isfile(fullpath) and f.startswith(installer_prefix):
+                keep_files.append(fullpath)
+
+    logger.info("Cleaning [%s] - keep_files:%s", installer_dir, keep_files)
+
+    # assume all files in file_list are in the same directory relative to dirname(installer)
+    try:
+        for file in os.listdir(installer_dir):
+            file = os.path.join(installer_dir, file)
+            if os.path.isfile(file) and file not in keep_files:
+                logger.info("Deleting file [%s] - not in keep_files", file)
+                os.remove(file)
+
+        # walk up and delete empty directories, but stop if the parent is one of the roots used by MG
+        # this is just maintenance to prevent aggregating empty directories in cache
+        remove_empty_dirs_upwards(installer_dir, installer_root_dirs)
+    except BaseException as e:
```

Comment:
```suggestion
    except Exception as e:
```
BaseException is probably a bit too broad.